syntax/parse.go

```diff
@@ -938,3 +950,103 @@ func terminatesExprList(tok Token) bool {
 	}
 	return false
 }
+
+// Comment assignment.
+// We build two lists of all subnodes, preorder and postorder.
+// The preorder list is ordered by start location, with outer nodes first.
+// The postorder list is ordered by end location, with outer nodes last.
+// We use the preorder list to assign each whole-line comment to the syntax
+// immediately following it, and we use the postorder list to assign each
+// end-of-line comment to the syntax immediately preceding it.
+
+// flattenAST returns the list of AST nodes, both in prefix order and in postfix
+// order.
+func flattenAST(root Node) (pre, post []Node) {
+	stack := []Node{}
+	Walk(root, func(n Node) bool {
+		if n != nil {
+			pre = append(pre, n)
+			stack = append(stack, n)
+		} else {
+			post = append(post, stack[len(stack)-1])
+			stack = stack[:len(stack)-1]
+		}
+		return true
+	})
+	return pre, post
+}
+
+// assignComments attaches comments to nearby syntax.
+func (p *parser) assignComments(n Node) {
+	pre, post := flattenAST(n)
+
+	// Split into whole-line comments and suffix comments.
+	var line, suffix []Comment
+	for _, com := range p.in.comments {
+		if com.Suffix {
+			suffix = append(suffix, com)
+		} else {
+			line = append(line, com)
+		}
+	}
+
+	// Assign line comments to syntax immediately following.
+	for _, x := range pre {
+		start, _ := x.Span()
+
+		switch x.(type) {
+		case *File:
+			continue
+		}
+
+		for len(line) > 0 && !start.isBefore(line[0].Start) {
+			x.AllocComments()
+			x.Comments().Before = append(x.Comments().Before, line[0])
+			line = line[1:]
+		}
+	}
+
+	// Remaining line comments go at end of file.
+	if len(line) > 0 {
+		n.AllocComments()
+		n.Comments().After = append(n.Comments().After, line...)
+	}
+
+	// Assign suffix comments to syntax immediately before.
+	for i := len(post) - 1; i >= 0; i-- {
+		x := post[i]
+
+		// Do not assign suffix comments to file
+		switch x.(type) {
+		case *File:
+			continue
+		}
+
+		_, end := x.Span()
+		for len(suffix) > 0 && end.isBefore(suffix[len(suffix)-1].Start) {
```

Comment:
It would be simpler and more efficient to find the least index for which the isBefore condition holds, then make a single call to AllocComments and append. You also wouldn't need to reverse the slice later.

Reply:
Still not done.

Reply:
Well, I've removed the reverse. If we assume there's at most one suffix comment, loop should be done 0 or 1 time. I don't think we can make it more efficient.

Reply:
We'll, we can avoid unnecessarily re-checking the loop condition: replace "for" with "if".